analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs

```diff
@@ -47,6 +47,9 @@ public abstract class SonarAnalysisContextBase<TContext> : SonarAnalysisContextB
     public SonarAnalysisContext AnalysisContext { get; }
     public TContext Context { get; }
 
+    private readonly ConcurrentDictionary<SyntaxTree, bool> shouldAnalyzeTreeCache = new();
```

Comment:
Please open another PR, where the `ConcurrentDictionary<SyntaxTree, bool>` is used the same way as `lastShouldAnalyze` in https://github.com/SonarSource/sonar-dotnet/pull/8407/files

* Caches the result of ShouldAnalyzeTree && LegacyIsRegisteredActionEnabled && ShouldAnalyzeRazorFile
* Is scoped by the capture in the CompilationStart
* Is scoped to RegisterNodeAction only

That way, we can compare the results between `lastShouldAnalyze` and the concurrent dict.